File: leetcode/Maximal_Rectangle.java
import java.util.*;

public class Solution {
    public int maximalRectangle(char[][] matrix) {
        int nrow = matrix.length;
        if (nrow == 0) return 0;
        int ncol = matrix[0].length;
        int[][] h = new int[nrow][ncol];
        for (int i = nrow - 1; i >= 0; --i)
            for (int j = 0; j < ncol; ++j) {
                if (matrix[i][j] == '0') {
                    h[i][j] = 0;
                } else {
                    if (i == nrow - 1 || matrix[i + 1][j] == '0') h[i][j] = 1;
                    else h[i][j] = h[i + 1][j] + 1;
                }
            }

        int ret = 0;
        int[] e = new int[nrow + 1];
        for (int i = 0; i < nrow; ++i) {
            int c = 0;
            e[0] = -1;
            for (int j = 1; j <= nrow - i; ++j) e[j] = -100;
            for (int j = 0; j < ncol; ++j) {
                while (c < h[i][j]) {
                    if (e[c] != - 100) {
                        int area = c * (j - e[c] + 1);
                        if (area > ret) ret = area;
                    }
                    ++c;
                    e[c] = j;
                }
                while (c > h[i][j]) {
                    e[c] = -100;
                    --c;
                    int area = c * (j - e[c]);
                    if (area > ret) ret = area;
                }
                int area = c * (j - e[c] + 1);
                if (area > ret) ret = area;
            }
            while (c > 0) {
                --c;
                int area = c * (ncol - e[c]);
                if (area > ret) ret = area;
            }

            c = 0;
            e[0] = ncol;
            for (int j = 1; j <= nrow - i; ++j) e[j] = -100;
            for (int j = ncol - 1; j >= 0; --j) {
                while (c < h[i][j]) {
                    if (e[c] != -100) {
                        int area = c * (e[c] - j + 1);
                        if (area > ret) ret = area;
                    }
                    ++c;
                    e[c] = j;
                }
                while (c > h[i][j]) {
                    e[c] = -100;
                    --c;
                    int area = c * (e[c] - j);
                    if (area > ret) ret = area;
                }
                int area = c * (e[c] - j + 1);
                if (area > ret) ret = area;
            }
            while (c > 0) {
                --c;
                int area = c * (e[c] + 1);
                if (area > ret) ret = area;
            }
        }
        return ret;
    }
}

File: leetcode/Maximal_Rectangle.cpp
class Solution {
public:
    // Input is at most 200x200
    int maximalRectangle(vector<vector<char> > &matrix) {
        int num_row = matrix.size();
        if (num_row == 0)
            return 0;
        int num_col = matrix[0].size();
        if (num_col == 0)
            return 0;    
        vector<vector<int> > rightmost(num_row, vector<int>(num_col));
        vector<vector<int> > lowest(num_row, vector<int>(num_col));
        for (int i = 0; i < num_row; ++i) {
            int rightmost_pos = -1;
            for (int j = num_col - 1; j >= 0; --j) {
                if (matrix[i][j] == '1') {
                    if (rightmost_pos == -1)
                        rightmost_pos = j;
                } else {
                    rightmost_pos = -1;
                }
                rightmost[i][j] = rightmost_pos;
            }
        }
        for (int j = 0; j < num_col; ++j) {
            int lowest_pos = -1;
            for (int i = num_row - 1; i >= 0; --i) {
                if (matrix[i][j] == '1') {
                    if (lowest_pos == -1)
                        lowest_pos = i;
                } else {
                    lowest_pos = -1;
                }
                lowest[i][j] = lowest_pos;
            }
        }
        int max_rect = 0;
        // O(N^3)
        for (int i = 0; i < num_row; ++i)
            for (int j = 0; j < num_col; ++j)
                if (matrix[i][j] == '1') {
                    int min_row_length = num_col;
                    for (int k = i; k <= lowest[i][j]; ++k) {
                        if (min_row_length > rightmost[k][j] - j + 1)
                            min_row_length = rightmost[k][j] - j + 1;
                        if (max_rect < (k - i + 1) * min_row_length)
                            max_rect = (k - i + 1) * min_row_length;
                    }
                }
        return max_rect;
    }
};
